[expo-updates] Update spec for version 1: message part (#20259)
expo/expo: docs/pages/technical-specs/expo-updates-1.mdx

> ---

Reviewer: If you are going to do this, you need to also modify `docs/constants/navigation.js` to navigate to `expo-updates-1` instead of `expo-updates-0`

Reviewer 2: We also will need to add redirect in:
* https://github.com/expo/expo/blob/main/docs/deploy.sh
* https://github.com/expo/expo/blob/main/docs/common/error-utilities.ts

An _update_ is defined as one or both of the following:
- a [_manifest_](#manifest-response) together with the assets referenced inside the manifest.
- a [_message_](#message-response) directive.

Expo Updates is a protocol for assembling and delivering updates to apps running on multiple platforms.

An app running a conformant Expo Updates client library MUST load the most recent _manifest_ saved in the client library's cache possibly after filtering by the contents of the manifest's [_metadata_](#manifest-body).

The following describes how a conformant Expo Updates client library MUST retrieve a new update from a conformant server:

1. The client library will make a [request](#update-request) for the most recent update, with constraints specified in the headers.
2. If a new update is downloaded, the client library will process its parts:
    - For a _manifest_ part, the client library will proceed to make additional requests to download and store any missing assets specified in the manifest. This is considered a new _manifest_.

>     - For a _message_ part, the client library will consume the message depending on the message type and update the its local state accordingly.

Reviewer: ```suggestion
    - For a _message_ part, the client library will consume the message depending on the message type and update its local state accordingly.
```

The primary consumers of this spec are Expo Application Services and organizations that wish to manage their own update server to satisfy internal requirements.

## Update Request

A conformant client library MUST make a GET request with the headers:

1. `expo-protocol-version: 1`, to specify version 1 of this expo-updates specicifcation.
2. `expo-platform`, to specify the platform type the client is running on.
   - iOS MUST be `expo-platform: ios`.
   - Android MUST be `expo-platform: android`.
   - If it is not one of these platforms, the server SHOULD return a 400 or a 404
3. `expo-runtime-version` MUST be a runtime version compatible with the client. A runtime version stipulates the native code setup a client is running. It should be set when the client is built. For example, in an iOS client, the value may be set in a plist file.
4. Any headers stipulated by a previous responses' [server defined headers](#manifest-response-headers).

> A conformant client library MUST send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` based on the supported response structures though SHOULD send `accept: application/expo+json, application/json, multipart/mixed`. A conformant client library MAY express preference using "q" parameters as specified in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1), which default to `1`.

Reviewer: Not sure what is meant by `based on the supported response structures though SHOULD send` -- can you clarify this?

Reviewer: Let's discuss this on Monday to make sure I understand it :)

Author: Essentially a compliant client can choose to only support non-multipart for simplicity (for example, if a consumer of the protocol only needs the manifest and doesn't need any of the other stuff). This sentence is saying that it has the option, but it is suggested to support multipart responses since there is more that can be done with those responses.

Conforming servers and client libraries must fulfill all normative requirements. Conformance requirements are described in this document by both descriptive assertions and key words with clearly defined meanings.

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in the normative portions of this document are to be interpreted as described in [IETF RFC 2119](https://tools.ietf.org/html/rfc2119). These key words may appear in lowercase and still retain their meaning unless explicitly declared as non-normative.

A conforming implementation of this protocol MAY provide additional functionality, but MUST NOT where explicitly disallowed or would otherwise result in non-conformance.

### Overview

Conforming servers and client libraries MUST follow the HTTP spec as described in [RFC 7231](https://tools.ietf.org/html/rfc7231) as well as the more precise guidance described in this spec.

> An _update_ is defined as one or both of the following:

Reviewer 3: As explained in the feedback summary, I think directives should not be considered updates. An update is an object the client app launches in order to run, hence `update = (manifest, assets)`

- a [_manifest_](#manifest-response) together with the assets referenced inside the manifest.
- a [_directive_](#directive-response).

> 

Reviewer 3: The spec goes on to talk about multipart responses but not directive responses -- probably need to fix something here.

Expo Updates is a protocol for assembling and delivering updates to apps running on multiple platforms.

> An app running a conformant Expo Updates client library MUST load the most recent _manifest_ saved in the client library's cache possibly after filtering by the contents of the manifest's [_metadata_](#manifest-body).

Reviewer 3: If a manifest has been saved but the entire update hasn't been fully downloaded, we don't want to load that manifest.

Separately (unrelated to this PR), "cache" is the wrong word here in that a cache implies the data can be arbitrarily erased and repopulated, but in our case we want to make sure that updates are stable.
```suggestion
An app running a conformant Expo Updates client library MUST load the most recent update saved in the client library's update database possibly after filtering by the contents of the manifest's [_metadata_](#manifest-body).
```

The primary consumers of this spec are Expo Application Services and organizations that wish to manage their own update server to satisfy internal requirements.

## Update Request

A conformant client library MUST make a GET request with the headers:

1. `expo-protocol-version: 1`, to specify version 1 of this expo-updates specicifcation.
2. `expo-platform`, to specify the platform type the client is running on.
   - iOS MUST be `expo-platform: ios`.
   - Android MUST be `expo-platform: android`.
   - If it is not one of these platforms, the server SHOULD return a 400 or a 404
3. `expo-runtime-version` MUST be a runtime version compatible with the client. A runtime version stipulates the native code setup a client is running. It should be set when the client is built. For example, in an iOS client, the value may be set in a plist file.
4. Any headers stipulated by a previous responses' [server defined headers](#manifest-response-headers).

> A conformant client library MUST send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` based on the [supported response structures](#update-response) though SHOULD send `accept: application/expo+json, application/json, multipart/mixed`. A conformant client library MAY express preference using "q" parameters as specified in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1), which default to `1`.

Reviewer: ```suggestion
A conformant client library MUST send at least one of `accept: application/expo+json`, `accept: application/json`, or `accept: multipart/mixed` based on the [supported response structures](#update-response), though it SHOULD send `accept: application/expo+json, application/json, multipart/mixed`. A conformant client library MAY express preference using "q" parameters as specified in [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.1), which default to `1`.
```

Example:

```
expo-protocol-version: 1
accept: application/expo+json;q=0.9, application/json;q=0.8, multipart/mixed
expo-platform: *
expo-runtime-version: *
expo-expect-signature: sig, keyid="root", alg="rsa-v1_5-sha256"
```

## Response

> A conformant server MUST return a response structured in at least one of the two following response structures, MAY support either or both response structures, and when an unsupported response structure is requested the server SHOULD respond with an HTTP `415` error status. A server that wishes to respond with an incompatible response for the requested protocol version SHOULD respond with an HTTP `404` error status instead.

Author: I went with 406 and 406 here since returning a 200 in this case might mean we'd want to return it in all cases and that seems like it might be a bigger change.

  - `sig` MUST contain the signature of the manifest. The name of this field matches that of `expo-expect-signature`.
  - `keyid` MAY contain the keyId of the key the server used to sign the response. The client SHOULD use the certificate that matches this `keyid` to verify the signature.
  - `alg` MAY contain the algorithm the server used to sign the response. The client SHOULD use this field only if it matches the algorithm defined for the certificate matching `keyid`.

### Multipart Response

A update response of this format is defined by the `multipart/mixed` MIME type as defined by [RFC 2046](https://tools.ietf.org/html/rfc2046#section-5.1).

Headers for this response format are the [common response headers](#common-response-headers), with the following exceptions:

- `content-type` SHOULD have a `multipart/mixed` value as defined by [RFC 2046](https://tools.ietf.org/html/rfc2046#section-5.1)

> Part order is not strict. A multipart response with no parts should be considered a no-op (no updates or directives available), though headers SHOULD be sent nevertheless.

Reviewer 3: Let's get this PR through but one thing we need to resolve in a follow-up is whether server-defined headers should be set when the server responds with an empty multipart response.If the client is expected to persist the server-defined headers, then no-op isn't the right word. 

We should take a look at expo-updates to see what it does today as well as our intent/motivation behind expo-server-defined-headers and then tidy up the spec. It's already kind of ambiguous about how to treat server-defined headers outside of the happy path where the server responds with a manifest.

- `assetRequestHeaders`: MAY contain a dictionary of header (key, value) pairs to include with asset requests. Key and value MUST both be strings.

### Directive Body

Defined as JSON conforming to both the following `Directive` definition expressed in [TypeScript](https://www.typescriptlang.org/) and the detailed descriptions for each field:

```typescript
type Directive = {

>   directiveType: string;
>   directivePayload?: { [key: string]: any };

Reviewer 3: Usually I like context-insensitive names in codebases but in a protocol the shorter names fit better IMO, what do you think?
```suggestion
  type: string;
  data?: { [key: string]: any };
```

>   [key: string]: any;

Reviewer 3: As I understand it, these extra fields are allowed for forwards-compatibility. The spec should say to allow these fields and to ignore any that aren't documented by the spec so that a client doesn't start relying on unspecced fields.

Author: These extra fields are for things like code signing context (scopeKey, projectId) and anything else that a client-server implementation may need. Maybe we should specify an `extra` field as well with the same style as `manifest.extra` (for `scopeKey`, `eas.projectId`). 

Also, more generally this spec says: 

> A conforming implementation of this protocol MAY provide additional functionality, but MUST NOT where explicitly disallowed or would otherwise result in non-conformance.

So I think I'll specify `extra` in the spec and remove the `[key: string]: any;`.

Reviewer 3: I do think it's a good idea to have the spec tell clients to allow and ignore unknown fields. It will help with forward compatibility if old clients will gracefully ignore fields instead of crashing if a new server sends back unknown fields.

};
```

- `directiveType`: The type of directive directive.
- `directivePayload`: MAY contain any extra information specific to the `directiveType`.

> A conformant client library and server SHOULD specify and implement directive types specific to the needs of the application. For example, Expo Application Services makes use of two directive types: `noUpdateAvailable` (directs the expo-updates library to proceed as normal) and `rollbackToEmbedded` (directs the expo-updates library to use the update embedded in the host application instead of any other downloaded updates).

Reviewer 3: Since all three multipart parts are optional, what if instead of a `noUpdateAvailable` directive, we had the spec treat an empty multipart response as a no-op? That would also save on response bytes. The client should also accept either a 204 or 200 in that case.

> A conformant client library and server SHOULD specify and implement directive types specific to the needs of the application. For example, Expo Application Services makes use of two directive types: `noUpdateAvailable` (directs the expo-updates library to proceed as normal) and `rollbackToEmbedded` (directs the expo-updates library to use the update embedded in the host application instead of any other downloaded updates).

Reviewer 3: Is the directive type meant to be a verb phrase (roll back) or a noun phrase (a rollback)?

```suggestion
A conformant client library and server SHOULD specify and implement directive types specific to the needs of the application. For example, Expo Application Services makes use of two directive types: `noUpdateAvailable` (directs the expo-updates library to proceed as normal) and `rollBackToEmbedded` (directs the expo-updates library to use the update embedded in the host application instead of any other downloaded updates).
```

- `assetRequestHeaders`: MAY contain a dictionary of header (key, value) pairs to include with asset requests. Key and value MUST both be strings.

### Directive Body

Defined as JSON conforming to both the following `Directive` definition expressed in [TypeScript](https://www.typescriptlang.org/) and the detailed descriptions for each field:

```typescript
type Directive = {
  type: string;

>   payload?: { [key: string]: any };

Reviewer 3: The word "payload" to me emphasizes that the data is being transported and is the body of a message. That's true here but these fields are more like parameters and arguments to the directive.

Some parts of the HTML spec (like Content Security Policies) use the terms "directive name" and "directive value". I think "value" is too generic and also seems a bit off when some of our update directives don't even need payloads/values/parameters.

Four words I considered are:
- parameters: I think this is best and adds a little precision by implying they parameterize the directive while also staying fairly generic and flexible
- data: OK but rather generic
- value: too generic and also implies a bit that it's necessary
- arguments: while maybe it's more technically correct than parameters, it's less intuitive than parameters and also is a reserved word in JS and can make coding a little less smooth
